subsys/bluetooth/services/ots/ots.c

```diff
@@ -271,63 +288,86 @@ static ssize_t ots_obj_prop_read(struct bt_conn *conn,
-	size_t name_len;
+	struct bt_gatt_ots_object *new_obj;
 
 	if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ) && ots->dir_list &&
 	    ots->dir_list->dir_list_obj->state.type != BT_GATT_OTS_OBJECT_IDLE_STATE) {
 		LOG_DBG("Directory Listing Object is being read");
 		return -EBUSY;
 	}
 
-	name_len = strlen(obj_init->name);
-
-	CHECKIF(name_len == 0 || name_len > CONFIG_BT_OTS_OBJ_MAX_NAME_LEN) {
-		LOG_DBG("Invalid name length %zu", name_len);
-		return -EINVAL;
-	}
-
-	CHECKIF(!ots_obj_validate_prop_against_oacp(obj_init->props, ots->features.oacp)) {
-		LOG_DBG("Object properties (0x%04X) are not a subset of OACP (0x%04X)",
-				obj_init->props, ots->features.oacp);
-		return -ENOTSUP;
-	}
-
-	err = bt_gatt_ots_obj_manager_obj_add(ots->obj_manager, &obj);
+	err = bt_gatt_ots_obj_manager_obj_add(ots->obj_manager, &new_obj);
 	if (err) {
 		LOG_ERR("No space available in the object manager");
 		return err;
 	}
 
-	/* Initialize object. */
-	memcpy(&obj->metadata, obj_init, sizeof(obj->metadata));
-
 	if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ)) {
-		bt_ots_dir_list_obj_add(ots->dir_list, ots->obj_manager, ots->cur_obj, obj);
+		bt_ots_dir_list_obj_add(ots->dir_list, ots->obj_manager, ots->cur_obj, new_obj);
 	}
 
-	/* Request object data. */
+	/* Initialize the object metadata */
+	memset(&new_obj->metadata, 0, sizeof(new_obj->metadata));
+
 	if (ots->cb->obj_created) {
-		err = ots->cb->obj_created(ots, NULL, obj->id, obj_init);
+		err = ots->cb->obj_created(ots, NULL,
+			new_obj->id, size, type,
+			&new_obj->metadata.name, &new_obj->metadata.size, &new_obj->metadata.props);
+
 		if (err) {
-			bt_gatt_ots_obj_manager_obj_delete(obj);
+			bt_gatt_ots_obj_manager_obj_delete(new_obj);
 
 			if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ)) {
 				bt_ots_dir_list_obj_remove(ots->dir_list, ots->obj_manager,
-							   ots->cur_obj, obj);
+							   ots->cur_obj, new_obj);
 			}
 
 			return err;
 		}
 	}
 
+	__ASSERT(ots_obj_validate_prop_against_oacp(new_obj->metadata.props, ots->features.oacp),
+		 "Object properties (0x%04X) are not a subset of OACP (0x%04X)",
+		 new_obj->metadata.props, ots->features.oacp);
+
+	__ASSERT(new_obj->metadata.name != NULL,
+		 "Object name must be set by application after object creation.");
+
+	new_obj->metadata.type = *type;
+
+	if (obj) {
+		*obj = new_obj;
+	}
+
 	return 0;
 }
 
+int bt_ots_obj_add(struct bt_ots *ots, uint32_t size, const struct bt_ots_obj_type *type)
+{
+	int err;
+	size_t name_len;
+	struct bt_gatt_ots_object *obj;
+
+	err = bt_ots_obj_add_internal(ots, NULL, size, type, &obj);
+	if (err) {
+		return err;
+	}
+
+	name_len = strlen(obj->metadata.name);
+	__ASSERT(name_len > 0 && name_len <= CONFIG_BT_OTS_OBJ_MAX_NAME_LEN,
+		"Invalid name length %zu", name_len);
+	__ASSERT(obj->metadata.size.alloc >= size,
+		 "Object allocated size must be greater than or equal to requested size.");
+	__ASSERT(obj->metadata.size.cur <= size,
+		 "Object current size must be less than or equal to requested size.");
```

Comment:
Should/could these be moved to where we have the __ASSERT for  `__ASSERT(new_obj->metadata.name != NULL`?